### PR TITLE
[CALCITE-5392] Support Snapshot in RelMdExpressionLineage

### PR DESCRIPTION
**Pull request by JiajunBernoulli**

### CODE REVIEW

**Review by asolimando — APPROVED** [review]
LGTM, just a minor comment

**Comment by asolimando** [comment]
@JiajunBernoulli, until the review process is over please don't force-push, it destroys the "link" to the old comments and it makes incremental reviewing impossible (checking just the delta from the latest review). In this case the PR is small-sized, but let's avoid it altogether

File: core/src/main/java/org/apache/calcite/rel/metadata/RelMdExpressionLineage.java [hunk 1]
```diff
@@ -391,6 +392,14 @@ protected RelMdExpressionLineage() {}
     return mq.getExpressionLineage(rel.getInput(), outputExpression);
   }
 
+  /**
+   * Expression lineage from Snapshot.
+   */
```
[the remark this answers is not in the document]
Comment (JiajunBernoulli) [reply core/src/main/java/org/apache/calcite/rel/metadata/RelMdExpressionLineage.java]: Thanks for your review, I added javadoc.
If you think that they is reasonable, I can add for other methods.

**Comment by JiajunBernoulli** [comment]
Sorry, I'll pay attention later.

File: core/src/main/java/org/apache/calcite/rel/metadata/RelMdExpressionLineage.java [hunk 2]
```diff
@@ -391,6 +392,18 @@ protected RelMdExpressionLineage() {}
     return mq.getExpressionLineage(rel.getInput(), outputExpression);
   }
 
+  /**
+   * Expression lineage from Snapshot.
+   * @param rel Relational expression which is Snapshot
+   * @param mq Metadata query
+   * @param outputExpression Expression which need be inferred
+   * @return If the lineage cannot be inferred, we return null.
+   */
```
Comment (asolimando) [inline core/src/main/java/org/apache/calcite/rel/metadata/RelMdExpressionLineage.java:401]: Thanks for adding the javadoc, in general the sentence starts with lower-case letters, I have taken the liberty to also rephrase a couple of sentences. No problem if you want to keep the original sentences, but please fix the first letter casing if you do so.

```suggestion
  /**
   * Expression lineage from Snapshot.
   * @param rel Snapshot relational expression
   * @param mq metadata query
   * @param outputExpression expression which needs to be inferred
   * @return the inferred lineage, possibly null.
   */
```

File: core/src/main/java/org/apache/calcite/rel/metadata/RelMdExpressionLineage.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (asolimando) [reply core/src/main/java/org/apache/calcite/rel/metadata/RelMdExpressionLineage.java]: It's better to keep the PR focused on the topic I guess, but thanks for volunteering

**Comment by asolimando** [comment]
@JiajunBernoulli it's good to me, if you can squash to a single commit I will merge, thanks!

**Comment by JiajunBernoulli** [comment]
@asolimando Thank you again, I  squashed to a single commit.

**Pull request merged**